### ISSUE TITLE
CFP: Support ClusterNetworkPolicy 

### ISSUE THREAD

**Issue by sathieu** [cilium/cilium#23380]
## Cilium Feature Proposal
**Is your feature request related to a problem?**

No

**Describe the feature you'd like**

Support for AdminNetworkPolicy, as described in https://github.com/kubernetes/enhancements/issues/2091

**Comment by pchaigno** [comment]
FYI, the difficult part here is the priority-based policies. It's generally difficult to build packet classification algorithms that take priorities into account while being efficient at scale.

**Comment by sathieu** [comment]
@pchaigno Thanks for the info. I don't know if Cilium was involved in defining the specs, and maybe something was missed at that time. I see that AdminNetworkPolicy (with BaselineAdminNetworkPolicy) is half of the future NPv2, the other being developper-focused policies (mabe by using current NetworkPolicies).

As cluster admins, we need cluster-scoped firewall rules (both baseline rules, overrridable ; and  non-overridable rules). Current [CiliumClusterwideNetworkPolicy](https://docs.cilium.io/en/stable/concepts/kubernetes/policy/#ciliumclusterwidenetworkpolicy) lacks some of our needs:

- rules are overridable by namespace-scoped resources
- when a pod is matched by a rule, the pod goes to default deny

Also we prefer to use non vendor-specific CRDs.

AdminNetworkPolicy (with BaselineAdminNetworkPolicy) match our needs, but the priority field is a very complex way to fulfill our needs (as mentioned [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy#risks-and-mitigation)).

**Comment by ysksuzuki** [comment]
I'm interested in the AdminNetworkPolicy support because we see the potential to make our policy setup simple and easy to maintain. I'm wondering if it would be possible to interpret the AdminNetworkPolicy manifests on the agent side and put the data in the policy map with the current data structure(or with some tweeks). Obviously, it's not that easy though. I think it's worth trying it.

**Comment by tssurya** [comment]
/cc

I work on the upstream ANP/BANP API: https://github.com/kubernetes-sigs/network-policy-api 
I am interested in this issue as we are looking to unify the way CNIs do the NP API's to align with the upstream one. Let us know if there are any major "can't implement or translate this x part of the API in our plugin" kind of roadblocks and please feel free to give us feedback on use cases we might not have considered. We are currently on the alphav1 version and want to understand the challenges implementors are running into before graduation.

I can see there is a note https://github.com/cilium/cilium/issues/23380#issuecomment-1423328510 on the difficultly of implementing the "priority" field.. Is that only from a point of view of efficiency or other aspects as well? Priority is actually a powerful way to classify the rules for the API specially since the scope is cluster-wide..

**Comment by sathieu** [comment]
not stale

NB: we are able to workaround the feature by using [`always`  Policy Enforcement Mode](https://docs.cilium.io/en/stable/security/policy/intro/#policy-enforcement-modes) and [Deny policies](https://docs.cilium.io/en/stable/security/policy/language/#id8). However, this is Cilium-specific.

**Comment by pchaigno** [comment]
Hi everyone,
This is the official GitHub "cc" button :smile_cat: :point_down: 
<img src="https://github.com/cilium/cilium/assets/1764210/63bd8434-dc97-496d-b84d-25329779f4cc" width="500">

**Comment by oblazek** [comment]
> NB: we are able to workaround the feature by using [always Policy Enforcement Mode](https://docs.cilium.io/en/stable/security/policy/intro/#policy-enforcement-modes) and [Deny policies](https://docs.cilium.io/en/stable/security/policy/language/#id8). However, this is Cilium-specific.

agree, we are forced to do the same via CCNPs and letting our users to create only CNPs

**Comment by squeed** [comment]
We are definitely looking in to this; the tricky bit is determining exactly how to implement ordering within the `toMapState()` empire. Specifically, the precedence within [denyPreferredInsert](https://github.com/cilium/cilium/blob/f1925b59247739ab3beb5ed3b395025d887c2825/pkg/policy/mapstate.go#L767)

**Comment by danehans** [comment]
> It will probably not happen for v1.16, but we are laying the ground work as part of port-range support.

@squeed do you have a link to any issues/PRs related for this work?

**Comment by bowei** [comment]
I'm interesting in helping out on this one.

I'm curious about the efficiency issue -- the use of priorities does (naively) require a chained implementation. I'm assuming g(enerally speaking), Cilium is looking for an O(1) implementation in the datapath?

**Comment by hemanthmalla** [comment]
@bowei Yes, thats my understanding based on some discussions in previous sig-policy meetings.

**Comment by pchaigno** [comment]
We used to have an O(1), hashmap-based implementation for the policy lookups (not including the identity resolution). But [it changed somewhat recently](https://github.com/cilium/cilium/pull/23885) and now uses LPM map lookups.

**Comment by tamilmani1989** [comment]
is there any design doc or proposal started for this feature?

**Comment by bowei** [comment]
I'm starting to collect info for a design doc.

**Comment by bowei** [comment]
For reference, I've started putting together some sketches to get the idea collection going. Would appreciate any suggestions/comments:

https://docs.google.com/document/d/1oO8K4deI4dHjpKWd0ICj0htUklT-2C0krBKjcGcR-CM/edit?tab=t.0

Happy to discuss this on `sig-policy` channel on Cilium Slack

**Comment by sathieu** [comment]
I've renamed this issue to ClusterNetworkPolicy.

It's supported in Cilium 1.20 ([ref](https://isovalent.com/blog/post/cilium-1-20/#kubernetes-clusternetworkpolicy-kcnp-support)), but [behind a feature-flag](https://docs.cilium.io/en/v1.20/network/kubernetes/policy/#kubernetes-clusternetworkpolicy) `--enable-k8s-cluster-network-policy`.

I don't know if this should be closed now, or once the flag is enabled by default.

**Comment by TheBeeZee** [comment]
I think it's fine to close this as completed. The main reason we didn't enable it by default in 1.20 is because it is an alpha API. The k8s sig-network-policy is hoping to release a v1beta1 of the API (with no actual API changes) soon-ish and I think that would be a good opportunity to make the feature enabled by default.

**Comment by squeed** [comment]
@TheBeeZee seems reasonable. Are you satisfied with the state of kCNP CI? Should we make a follow-up issue to update CI when v1beta1 cuts?

**Comment by TheBeeZee** [comment]
> [@TheBeeZee](https://github.com/TheBeeZee) seems reasonable. Are you satisfied with the state of kCNP CI? Should we make a follow-up issue to update CI when v1beta1 cuts?

There is still a number of things I want to complete:
* Implement flags for enabling FQDN policy (in progress)
* Add the ability to auto-install the CRD
* Improve CI
* Enable KCNP by default
* Update to v1beta1 (when released)

Unless you disagree, this CFP is likely not the right issue to track this.
